Introduce virtual TensorImpl::layout_impl() and sparse CSC/BSR/BSC layouts

pearu opened the pull request:
Replaced by https://github.com/pytorch/pytorch/pull/75596

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75585
* __->__ #75023
* #75594

pearu committed:
Introduce sparse CSC layout and is_sparse_csc property - 2.

[ghstack-poisoned]

pearu committed:
Update on "Introduce sparse CSC layout and is_sparse_csc property - 2."

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce sparse CSC layout and is_sparse_csc property - 2."

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

cpuhrsch commented on c10/core/TensorImpl.h:
```diff
@@ -2638,6 +2651,14 @@ struct C10_API TensorImpl : public c10::intrusive_ptr_target {
+  // tensor attributes, set define_layout_from_dispatch_key to
+  // false. In this case, the corresponding tensor impl class must
+  // overwrite virtual layout_impl() method.
+  bool define_layout_from_dispatch_key_ : 1;
```
We don't need this extra bit though, at least for now. It's more general, but it'll impose additional memory cost.

pearu replied:
Removed.

cpuhrsch commented on tools/autograd/gen_python_functions.py:
```diff
@@ -70,7 +70,7 @@
 # These functions require manual Python bindings or are not exposed to Python
 _SKIP_PYTHON_BINDINGS = [
-    'alias', 'contiguous', 'is_cuda', 'is_sparse', 'is_sparse_csr', 'size', 'stride',
+    'alias', 'contiguous', 'is_cuda', 'is_sparse', 'is_sparse_csr', 'is_sparse_csc', 'size', 'stride',
```
nit: forgot to remove this?

cpuhrsch replied:
I assume this is just WIP

pearu replied:
Yes, it is gone now.

pearu committed:
Update on "Introduce sparse CSC layout and virtual TensorImpl::layout_impl()."

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

cpuhrsch commented on aten/src/ATen/core/tensor_type.cpp:
```diff
@@ -253,7 +253,7 @@ TensorTypePtr TensorType::create(const at::Tensor& t) {
   VaryingShape<int64_t> strides;
   VaryingShape<int64_t> sizes;
-  if (!t.is_mkldnn() && !t.is_sparse() && !t.is_sparse_csr()) {
+  if (!t.is_mkldnn() && !t.is_sparse() && !t.is_sparse_csr() && t.layout() != at::kSparseCsc) {
```
Same here

pearu replied:
Fixed.

pearu committed:
Update on "Introduce sparse CSC layout and virtual TensorImpl::layout_impl()."

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce sparse CSC layout and virtual TensorImpl::layout_impl()."

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce sparse CSC layout and virtual TensorImpl::layout_impl()."

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

albanD commented:
I'll let @ezyang review this one as he has a much better understanding of the constraints here.

pearu committed:
Update on "Introduce sparse CSC layout and virtual TensorImpl::layout_impl()."

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC|BSR|BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC|BSR|BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC|BSR|BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC|BSR|BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC|BSR|BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC|BSR|BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC|BSR|BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

ezyang commented on aten/src/ATen/SparseCsrTensorImpl.cpp:
```diff
@@ -23,6 +23,7 @@ DeviceType SparseCsrTensorSetToDeviceType(DispatchKeySet key_set) {
 
 SparseCsrTensorImpl::SparseCsrTensorImpl(
     at::DispatchKeySet key_set,
+    Layout layout,
```
Why do you have to take this as an argument, the layout is guaranteed to be CSR here no?

ezyang replied:
Answer: because it can also be BSR. Consider adding an assert here.

ezyang replied:
Maybe due for a rename of SparseCsrTensorImpl itself

pearu replied:
No, with this PR, SparseCsrTensorImpl can hold four layouts: kSparseCsr, kSparseCsc, kSparseBsr, and kSparseBsc

pearu replied:
Renaming SparseCsrTensorImpl makes a lot of sense. However, since this PR implements a new approach (dispatch key-layout relation is not many-to-one anymore), I'd suggest keeping the PR minimal for reviewers and doing the rename in a follow-up PR (which would be simple but very verbose).

ezyang commented on aten/src/ATen/SparseCsrTensorImpl.cpp:
```diff
@@ -44,24 +45,31 @@ SparseCsrTensorImpl::SparseCsrTensorImpl(
 void SparseCsrTensorImpl::resize_(int64_t nnz, IntArrayRef size) {
-  auto rows = size[0];
-  auto cols = size[1];
+  TORCH_CHECK((layout_ == kSparseCsr || layout_ == kSparseBsr), "resize_: layout ", layout_, " is not yet supported");  // TODO
```
With the assert in constructor you don't have to repeatedly test that the layout makes sense in these implementations

pearu replied:
We want to support the kSparseCsc and kSparseBsc as soon as possible. This check is to ensure that we don't forget to update this method because the code that follows would be wrong for CSC and BSC.

ezyang commented on aten/src/ATen/core/TensorBase.h:
```diff
@@ -244,7 +244,7 @@ class TORCH_API TensorBase {
     // Setting channels_last_strides_exact_match to true forces function to
     // check 0,1 - sized dimension strides.
-    if (!is_mkldnn() && !is_sparse()) {
+    if (layout() == at::kStrided) {
```
good

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -285,18 +371,29 @@ Tensor& copy_sparse_csr_(Tensor& self, const Tensor& src, bool non_blocking) {
-      self.toString(),
-      " and src type = ",
-      src.toString());
+  switch (src.layout()) {
```
I'd write a helper function for this pattern. It seems to come up a lot.

pearu replied:
Yes, this is a common pattern. However, the content of `case` blocks varies in different places and I am not sure what should be the corresponding helper function. A function that takes four lambda functions as arguments? So, can you be more specific on the helper function idea?

pearu replied:
Resolving as a no-op: the switch statement is the optimal approach.

cpuhrsch replied:
We have https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Dispatch.h to write generic code for each case, we could add something similar for layouts.

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -37,48 +37,74 @@ using namespace at::sparse_csr;
 
 namespace {
 
+Layout get_layout_from_members(const Tensor& crow_indices, const Tensor& col_indices, const Tensor& values) {
```
crow and col indices are unused, why pass them as argument?

cpuhrsch commented on aten/src/ATen/native/sparse/SparseCsrTensor.cpp:
```diff
@@ -157,17 +190,17 @@ SparseCsrTensor new_csr_tensor(const TensorOptions& options) {
     IntArrayRef size,
     c10::optional<ScalarType> dtype,
-    c10::optional<Layout> layout,
+    c10::optional<Layout> layout,  // unsafe requires explicit layout
```
This could cause serialization errors when loading old models https://github.com/pytorch/pytorch/blob/30653d164d7223263e41c31060f75ad0bf9d9cab/torch/_utils.py#L187-L194
This call site isn't passed a layout.

pearu replied:
`layout` is now a part of the tensor data and it has the same level of importance as `crow_indices`, `col_indices`, `values`, and `size`. It is specified as optional for BC.

Notice that the content of `_rebuild_sparse_csr_tensor` now contains:
```python
result = torch._sparse_csr_tensor_unsafe(crow_indices, col_indices, values, size, layout=torch.sparse_csr)
```
that is, loading old models are ok.

However, we might have a problem with loading new models, let me check...

pearu replied:
Fixed.

ezyang commented:
LGTM for the core parts. I did not review the sparse TensorImpl itself.

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC|BSR|BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC|BSR|BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC|BSR|BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC/BSR/BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC/BSR/BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC/BSR/BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC/BSR/BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu committed:
Update on "Introduce virtual TensorImpl::layout_impl() and sparse CSC/BSR/BSC layouts"

Replaces https://github.com/pytorch/pytorch/pull/74546 . Fixes https://github.com/pytorch/pytorch/issues/74916 .




[ghstack-poisoned]

pearu commented:
Retiring this PR in favor of https://github.com/pytorch/pytorch/pull/75596 to fix a weird failure
```
<snip>
Environment variable + a5c652cb5f9e330e8ef5b061423e94cfc017f2c2 Update base for Update on "Introduce sparse CSC layout and virtual TensorImpl::layout_impl()." not defined
'BSR' is not recognized as an internal or external command,
operable program or batch file.
```
in https://github.com/pytorch/pytorch/runs/5968822325?check_suite_focus=true

pearu commented:
Replaced by https://github.com/pytorch/pytorch/pull/75821
